### ISSUE TITLE
Replace use of term "Minion"

### ISSUE THREAD

**Issue by jbeda** [kubernetes/kubernetes#1111]
There has been quite a bit of confusion over our use of the term "minion".  While it captures what a worker node in a k8s cluster is well, it is highly associated with Salt.  This has led to some confusion where people think that Salt is playing a larger role in k8s than it actually is.

I propose that we change from Minion to something that is more Kubernetes specific.

Questions:
- Is this worthwhile?  Should we do this?
- What will this impact?  I know that minion has made it into our API so that is something we'll have to address.
- What term should we use to replace "Minion".  Personally, I'd like to avoid any term that around "slave".  Suggestions:
  - KNode
  - Koupi (greek for oar)
  - Rower
  - Krew
  - ???

**Comment by bgrant0607** [comment]
We're part way through this transition, leaving the system in a confusing state. We should finish cleaning this up.

**Comment by bgrant0607** [comment]
Ugh, this is still a mess:
```
grep -r -i minion docs pkg plugin |wc -l
927
```

**Comment by bgrant0607** [comment]
Still a confusing mix of node and minion. Uses of minion include (not exhaustive):
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/registry/minion/registry.go
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/registry/minion/etcd/etcd.go
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/registry/registrytest/minion.go
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/registry/pod/etcd/etcd_test.go
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/latest/latest.go
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/v1beta3/conversion_test.go
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/plugin/pkg/scheduler/algorithm/listers.go
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/plugin/pkg/scheduler/generic_scheduler.go
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/plugin/pkg/scheduler/algorithm/priorities/spreading.go
(lots of other hits in the scheduler)
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/contrib/completions/bash/kubectl

Not to mention all the uses in our cluster turnup scripts.

**Comment by thockin** [comment]
Honestly, just replacing it in the turnups would be a big deal - when I run
get nodes and see "minion" I cringe.

**Comment by bgrant0607** [comment]
Minion vs. node vs. worker vs. kubelet has come up as a point of confusion.

**Comment by ghodss** [comment]
Where does worker come from?

**Comment by bgrant0607** [comment]
```
docs/getting-started-guides/docker-multinode/worker.md
docs/getting-started-guides/aws-coreos.md:### Launch 3 worker nodes
docs/getting-started-guides/coreos/coreos_multinode_cluster.md:* Provision one or more worker nodes 
...
```

**Comment by bgrant0607** [comment]
And then there's host...

**Comment by thockin-cc** [comment]
Having started the migration from minion to node, it is clear that
migrating in either direction is significant work. We should pick the word
we like and then never look back.

**Comment by bgrant0607** [comment]
If we eliminated the term "master" and the concept of a master machine, does that change our preference?
```
grep -r -i master cluster cmd contrib docs examples hack pkg plugin test www |wc -l
2307
```

The future is hosting of the cluster-level control plane on Kubernetes (self-hosting and remote hosting): HA, plugins, custom schedulers, addons, ...

The terms aren't as catchy, but the following makes sense to me:
* cluster (software-defined -- what we call "cell" internally)
* API server
* control plane (incl. API server, store, schedulers, controllers, plugins, addons, etc.)
* [cluster] nodes

**Comment by smarterclayton** [comment]
No opinion about upstream, but in our documentation, tutorials, and education sessions we have to have a term for the control plane, and people expect a word they can relate to.  "Master" fits that for us today (replacing "broker" from previous versions).

**Comment by davidopp** [comment]
Regarding @bgrant0607's point, I think that if all nodes are going to be truly identical in the long run, then "nodes" makes sense. But if there are going to be *any* special nodes (maybe a special security policy or firewall for the API server, or special PD mounts for the etcd replicas) for the purpose of running any portion of the control plane, then I think it is worth distinguishing "minions" from the special nodes.

I think the other terms from Brian's list, like cluster and control plane, are also useful, but I think this is somewhat orthogonal to the question of what to call the "[cluster] nodes."

In other words, I vote for "minions."

**Comment by thockin** [comment]
*Someone* has to have permissions to access cloud-provider stuff like
attaching PDs.

On Thu, Jun 4, 2015 at 10:52 PM, David Oppenheimer <notifications@github.com

**Comment by smarterclayton** [comment]
In our docs we are starting to use "nodes" and "infrastructure nodes" or "infra" nodes to describe the difference.

**Comment by bgrant0607** [comment]
@thockin "Someone" doesn't have to be a node. It just needs to be a process with the right credentials.

**Comment by bgrant0607** [comment]
From #9176:
```
grep -r -i minion cluster cmd contrib docs examples hack pkg plugin test www |wc -l
1713
grep -r -i node cluster cmd contrib docs examples hack pkg plugin test www |wc -l
5411
```

**Comment by thockin** [comment]
And yet I feel like changing node to minion is less risky than the other
direction (except API stuff)

**Comment by bgrant0607** [comment]
Why less risky?

"Minion" (and "pod" for that matter) is fairly incongruous with the rest of the API design, which uses fairly prosaic terminology rather than "cute" terms. The issue of cute terms also came up in the usability study -- what's a "pod"?

**Comment by thockin** [comment]
I think it is somewhat less risky because "minion" has deeper roots.
Either way, we just need to pick and proceed carefully.  Since all the
momentum is behind node at this point, that's probably the direction that
makes most sense, even if minion feels more context-specific.

**Comment by davidopp** [comment]
I'm fine with "node" instead of "minion" but it *has* to be preceded by another word. Like "slave node" or "worker node" or something like that. The problem isn't the word "node" per se -- it's the word "node" by itself, which doesnt allow one to distinguish.

**Comment by ghodss** [comment]
That kind of requirement is rarely adhered to in practice. People will shorten it to node most of the time. Node/infra node is probably the right split. But as @bgrant0607 says, those two types will become more and more homogenous over time.

**Comment by ghodss** [comment]
In case we're voting again, I do like minion more - it's more clearly "a miner who gets work done" instead of the more vague "a component of a system". It will be quite painful to change the v1 API though.

**Comment by davidopp** [comment]
> That kind of requirement is hard to adhere to in practice. People will shorten it to node most of the time. 

I agree. I didn't mean to imply I was switching my vote away from "minion" just that I saw maybe a way we could still leave "node" in the term (but I think it is worse for the reason you mentioned).

**Comment by evasnee** [comment]
Chiming in on this to say this inconsistency caused confusion in our latest hackathon because we use three different terms (Node, Minion, Worker) throughout our documentation and in other forums where we discuss this concept.

![image](https://cloud.githubusercontent.com/assets/11644912/8043016/f9f50cba-0dd7-11e5-9a72-6837833bc51d.png)

In a recent OpenStack Murano presentation (https://www.youtube.com/watch?v=ao8UAShNBW8&feature=youtu.be&t=20m30s), the OpenStack UI and the presenter both refer to it as “minions.” If we keep using the word in our own documentation, we can’t expect non-Googlers to stop using the term.

Agree with @thockin that we need to choose one and proceed carefully.

**Comment by jiangyaoguo** [comment]
I hadn't seen this issue before I made this pr #10843. Do we have a conclusion about this issue now? or I should change "get minions" to "get nodes" only? @bgrant0607  @thockin

**Comment by bgrant0607** [comment]
Another observation:

We use Host in a number of places in the API: HostPort, HostIP, HostNetwork, HostPath, Host (in HTTPGetAction and EventSource), and Hostname. More often than Node (not counting the Node API itself): NodeSelector, NodeName, NodePort. "Host" appears to be more natural/familiar.

The main disadvantage of "Host" would be confusion about whether "HostName" referred to the resource name or DNS name.

**Comment by davidopp** [comment]
@bgrant0607 Host vs. node isn't really the issue in my mind. The reason I am advocating for "minion" is to distinguish nodes running user containers from nodes running the core (non-addon) master components. Even if we go to self-hosting I think most users will still want the master nodes to be special from a security policy standpoint if not for other reasons as well, for the duration that they are running the master components. And in GKE there is clearly a distinction because the master nodes are not even in the user's GCE project while the minions are.

**Comment by bgrant0607** [comment]
@davidopp You're saying that the nodes would be the same, except for policy. 

Even before we self-host, we're planning to register the master node(s) as nodes.

Fill in the blank: cluster of _____________.

I'm sorry, but minion is out of the running, at least as an API keyword.

**Comment by davidopp** [comment]
Certainly in the cases of fields where the thing being described could be a master node or a minion node, it would not make sense to call it a minion -- it should be "node" or "host" or whatever. I'm just talking about the situations where the thing being described is always a non-master node. Looking back at the first entry in this issue, I thought that's what this discussion was about, for example one of the rejected alternatives there was "slave."

**Comment by zhengguoyong** [comment]
hi, all
is there anynoe doing this, if no, i think i can make some PR, as i was doing this recently.
or you can make me assignee. thanks.
@jbeda  @bgrant0607 @thockin

**Comment by thockin** [comment]
There have been a few PRs in this direction, but mostly hitting the easy
stuff.  The big one will be when nodes stop coming up named as "minion"

**Comment by eosrei** [comment]
My above referenced PR includes ~400 of the ~1200 instances of "minion" in the codebase in three  commits. I started with the environment variables, kube-util helper functions, and Vagrant provider. Not sure if I should expand this one PR to include all ~1200 uses or break it apart into multiple PRs. Review and/or thoughts appreciated.

**Comment by thockin** [comment]
We're so close on this.  The right time to try to get the last (hardest) uses of "minion" is probably right now - just after a release.  This will get us the most testing and leniency for mistakes :)

Interested in finishing?

**Comment by jbeda** [comment]
@thockin -- who was the "interested in finishing" pointed at?

**Comment by eosrei** [comment]
@jbeda Good chance it was me.

@thockin I want to, but I'm in the middle of a few other projects and can't right now.

**Comment by bgrant0607** [comment]
Minion is still present in scripts in /cluster, such as:
https://github.com/kubernetes/kubernetes/blob/8b8fb6c70da3217a07f198b43528bdf16af4f883/cluster/gce/config-default.sh#L78

We should at least be able to eradicate it when we fully remove provider-specific code from this repo.

**Comment by andrey01** [comment]
Not sure if this is the right place and time, but how about having the ``master`` and a ``worker`` types of the node? (where node is a single system in a cluster of many systems)

I think that would remove one level of ambiguity.

**Comment by macalinao** [comment]
Sad! I love minions.

**Comment by thockin-cc** [comment]
I really want to see this finished!!

**Comment by bgrant0607** [comment]
https://github.com/kubernetes/kubernetes/search?utf8=%E2%9C%93&q=minion&type=

This only exists in a few places in the cluster directory, which we're going to eliminate at some point, so this done enough
